### PR TITLE
Gpu swap integration

### PR DESCRIPTION
**Pull request by fegin**
Change MemoryManager to be an abstract interface. Provide GetMemoryManager() and GetMemoryManagerRef() for upper layers to transparently access MemoryManager. The default MemoryManager is CUDAMemoryManager.

### CODE REVIEW

**Comment by fegin** [comment]
@Sotskin @KarlLeell Please check this pull request. Before doing so, please read my email.

File: include/mxnet/buddy.h [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+  if (size <= 128) return 0;
+  return ceil(log2(static_cast<double>(size)) - log2(static_cast<double>(MIN_ALLOC_SIZE)));
+}
+
+inline unsigned long GetListSize(size_t size) {
+  return GetListIdx(size) + 1;
+}
+
+inline size_t GetListBlockSize(int idx) {
+  return pow(2, idx + log2(static_cast<double>(MIN_ALLOC_SIZE)));
```
Comment (Sotskin) [inline include/mxnet/buddy.h:35]: pow is very slow. I suggest using left shift instead

File: src/storage/mem_mgr.cc [hunk 2]
```diff
@@ -1,32 +1,240 @@
...
+  std::cout << "Destroy Cuda Memory Allocator" << std::endl;
+}
+
+cudaError_t CudaMemoryManager::Malloc(void*& devptr, size_t size,
+                                      int device_id) {
+  cudaSetDevice(device_id);
+  cudaError_t e = cudaMalloc(&devptr, size);
+  if (e != cudaSuccess && e != cudaErrorCudartUnloading) {
+    std::cout << "Malloc failed: " << cudaGetErrorString(e) << std::endl;
+  }
```
Comment (Sotskin) [inline src/storage/mem_mgr.cc]: We can replace all these with CUDA_CALL now

File: include/mxnet/mem_mgr.h [hunk 3]
```diff
@@ -1,91 +1,86 @@
...
 typedef enum {
   memStatus_Sucess,
   memStatus_InvalidValue,
   memStatus_OutOfMemory,
-  memStatus_CUDAError,
+  memStatus_CUDAError
 } memStatus_t;
 
-inline char* memGetStatusString(memStatus_t status) {
+inline std::string MemGetStatusString(memStatus_t status) {
```
Comment (Sotskin) [inline include/mxnet/mem_mgr.h]: It seems nobody is using this function. Shall we remove it?

**Review by Sotskin — COMMENTED** [review]
Just some nits.

File: include/mxnet/buddy.h (hunk 1 above)
Comment (fegin) [reply include/mxnet/buddy.h:35]: I have not yet dived into buddy system. I'll do that in another branch. But I'll add a FIXME.

File: include/mxnet/buddy.h [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+namespace mxnet {
+
+typedef enum {
+  blockStatus_Uninitialized,
+  blockStatus_Free,
+  blockStatus_Allocated
+} blockStatus_t;
+  
+const int MIN_ALLOC_SIZE = 128;
+const size_t CLEAN_UP_BOUNDRY = 500000000;
```
Comment (KarlLeell) [inline include/mxnet/buddy.h]: It is fine for now, but I feel we should choose these boundaries dynamically in the future.

File: src/storage/mem_mgr.cc [hunk 5]
```diff
@@ -1,32 +1,240 @@
...
+  return e;
+}
 
-MemoryManager::MemoryManager() {
-  std::cout << "Initialize Memory Allocator" << std::endl;
+cudaError_t CudaMemoryManager::Memcpy(int device_id, void* dst, const void* src,
+                                      size_t count, enum cudaMemcpyKind kind) {
+  cudaSetDevice(device_id);
+  cudaSetDevice(device_id);
```
Comment (KarlLeell) [inline src/storage/mem_mgr.cc]: A duplicated cudaSetDevice()?

**Comment by Sotskin** [comment]
Another suggestion: comment out line 86 of gpu_swap_prefetch.cc, which call PrintHandles of swap function. For testing I think printing handles is unnecessary and takes time

**Review by Sotskin — CHANGES_REQUESTED** [review]
At line 23 of run.py, num_loop is not parsed, so the model will always run 30 loops. (Since this part is not changed in this pullreq I can't quote it like other reviews)

**Pull request merged**